staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go

```diff
@@ -96,60 +104,188 @@ func TestExplainNotExistResource(t *testing.T) {
-		t.Fatalf("expected output should include pod kind")
-	}
+	// We run each test case under a different profile with different options
+	openapiv2Name := "OpenAPIV2"
+	openapiv3Name := "OpenAPIV3"
 
-	cmd.Flags().Set("api-version", "batch/v1")
-	cmd.Run(cmd, []string{"cronjobs"})
-	if !strings.Contains(buf.String(), "VERSION:  batch/v1") {
-		t.Fatalf("expected output should include pod batch/v1")
+	cases := []explainTestCase{
+		{
+			Name:          "Basic",
+			Args:          []string{"pods"},
+			ExpectPattern: []string{`\s*KIND:[\t ]*Pod\s*`},
+		},
+		{
+			Name:          "Recursive",
+			Args:          []string{"pods"},
+			Flags:         map[string]string{"recursive": "true"},
+			ExpectPattern: []string{`\s*KIND:[\t ]*Pod\s*`},
+		},
+		{
+			Name:          "DefaultAPIVersion",
+			Args:          []string{"horizontalpodautoscalers"},
+			Flags:         map[string]string{"api-version": "autoscaling/v1"},
+			ExpectPattern: []string{`\s*VERSION:[\t ]*(v1|autoscaling/v1)\s*`},
+		},
+		{
+			Name:              "NonDefaultAPIVersion",
+			Args:              []string{"horizontalpodautoscalers"},
+			Flags:             map[string]string{"api-version": "autoscaling/v2"},
+			ExpectPattern:     []string{`\s*VERSION:[\t ]*(v2|autoscaling/v2)\s*`},
+			VersionsWhitelist: sets.New(openapiv3Name),
+		},
+		{
+			Name:               "NonExistingAPIVersion",
+			Args:               []string{"pods"},
+			Flags:              map[string]string{"api-version": "v99"},
+			ExpectErrorPattern: `couldn't find resource for \"/v99, (Kind=Pod|Resource=pods)\"`,
+		},
+		{
+			Name:               "NonExistingResource",
+			Args:               []string{"foo"},
+			ExpectErrorPattern: `the server doesn't have a resource type "foo"`,
+		},
+		{
+			// No --output, but OpenAPIV3 enabled should fall back to v2 if
+			// v2 is not available. Shows this by making openapiv3 client
+			// point to a bad URL. So the fact the proper data renders is
+			// indication v2 was used instead.
+			Name:              "Fallback",
+			Args:              []string{"pods"},
+			ExpectPattern:     []string{`\s*KIND:[\t ]*Pod\s*`},
+			OpenAPIV3SchemaFn: fallbackV3SchemaFn,
+			VersionsWhitelist: sets.New(openapiv3Name),
+		},
+		{
+			// Show that explicitly specifying --output plaintext-openapiv2 causes
+			// old implementation to be used even though OpenAPIV3 is enabled
+			Name:              "OutputPlaintextV2",
+			Args:              []string{"pods"},
+			Flags:             map[string]string{"output": "plaintext-openapiv2"},
+			ExpectPattern:     []string{`\s*KIND:[\t ]*Pod\s*`},
+			OpenAPIV3SchemaFn: fallbackV3SchemaFn,
+			VersionsWhitelist: sets.New(openapiv3Name),
+		},
 	}
 
-	cmd.Flags().Set("api-version", "batch/v1beta1")
-	cmd.Run(cmd, []string{"cronjobs"})
-	if !strings.Contains(buf.String(), "VERSION:  batch/v1beta1") {
-		t.Fatalf("expected output should include pod batch/v1beta1")
+	type catchFatal error
+
+	for _, vers := range []string{openapiv2Name, openapiv3Name} {
```

Comment:
Instead boiling openapiv2 and openapiv3 tests down to the one function, I'd suggest to run them in their own functions. For example, when openapiv2 is deprecated, this function will be changed fundamentally again(which hurts confidence due to making changes both in code and unit tests.). Moreover, there won't be any need to this loop and `VersionWhiteList` field.

Reply:
good point. refactored into separate tests and removed field